core/src/qos_service.rs

```diff
@@ -100,24 +113,31 @@ impl QosService {
         let mut cost_tracker = bank.write_cost_tracker().unwrap();
         let select_results = transactions
             .zip(transactions_costs)
-            .map(|(tx, cost)| match cost_tracker.try_add(tx, cost) {
-                Ok(current_block_cost) => {
-                    debug!("slot {:?}, transaction {:?}, cost {:?}, fit into current block, current block cost {}", bank.slot(), tx, cost, current_block_cost);
-                    self.metrics.selected_txs_count.fetch_add(1, Ordering::Relaxed);
-                    Ok(())
-                },
-                Err(e) => {
-                    debug!("slot {:?}, transaction {:?}, cost {:?}, not fit into current block, '{:?}'", bank.slot(), tx, cost, e);
-                    match e {
-                        CostTrackerError::WouldExceedBlockMaxLimit => {
-                            self.metrics.retried_txs_per_block_limit_count.fetch_add(1, Ordering::Relaxed);
-                            Err(TransactionError::WouldExceedMaxBlockCostLimit)
-                        }
-                        CostTrackerError::WouldExceedAccountMaxLimit => {
-                            self.metrics.retried_txs_per_account_limit_count.fetch_add(1, Ordering::Relaxed);
-                            Err(TransactionError::WouldExceedMaxAccountCostLimit)
+            .map(|(tx, cost)| {
+                if Self::fee_payer_has_enough_balance(tx, cost, bank) {
```

Comment:
decision of if a tx should be included in current block: 
1. if it's fee payer has enough lamports, otherwise drop this tx at once (no retries); 
2. if it's cost (with cost_weight considered) would fit into current block, if not, it'd pushed into retry queue